### PR TITLE
extend scd4x

### PR DESCRIPTION
**Pull request by martgras**
# What does this implement/fix?

- add slow periodic measurement mode to reduce power consumption.
- add single shot measurement (scd41 only)
- add forced calibration action
- add factory reset action

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2032

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
  - platform: scd4x
    id: scd40
    i2c_id: bus_a
    altitude_compensation: 418m
    automatic_self_calibration: true
    measurement_mode: low_power_periodic
    co2:
      name: SCD40 Co²
      id: co2
    temperature:
      name: "SCD40 Temperature"
      id: scdD41_temperature
      accuracy_decimals: 1
    humidity:
      name: "SCD40 Humidity"
      id: scd40_humidity
      accuracy_decimals: 0
    update_interval: ${updates}

button:
  - platform: template
    name: "Calibrate CO2"
    on_press:
      - scd4x.perform_forced_calibration:
          #value: 419   # outside average April 2022
          value: !lambda "{ return 419; }"
          id: scd40
  - platform: template
    name: "reset CO2"
    on_press:
      - scd4x.factory_reset: scd40

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/scd4x/sensor.py [hunk 1]
```diff
@@ -106,3 +130,45 @@ async def to_code(config):
...
+)
+async def scd4x_frc_to_code(config, action_id, template_arg, args):
+    var = cg.new_Pvariable(action_id, template_arg)
+    await cg.register_parented(var, config[CONF_ID])
+    value = config[CONF_VALUE]
+    if cg.is_template(value):
+        templ = await cg.templatable(value, args, cg.uint16)
+        cg.add(var.set_value_template(templ))
+    else:
+        cg.add(var.set_value_static(value))
```
Comment (jesserockz) [inline esphome/components/scd4x/sensor.py]: Why do this when the action can handle templatable directly?
See `TEMPLATABLE_VALUE` macro

Comment (martgras) [reply esphome/components/scd4x/sensor.py]: Oh nice - didn't know

File: esphome/components/scd4x/automation.h [hunk 2]
```diff
@@ -9,28 +9,16 @@ namespace scd4x {
...
  public:
-  void set_value_template(std::function<uint16_t(Ts...)> func) {
-    this->value_func_ = func;
-    this->static_ = false;
-  }
-  void set_value_static(uint16_t value) {
-    this->value_static_ = value;
-    this->static_ = true;
-  }
+  void set_value(uint16_t value) { this->value_ = value; }
```
Comment (jesserockz) [inline esphome/components/scd4x/automation.h]: The macro takes care of this too.
```suggestion
```

File: esphome/components/scd4x/automation.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    if (this->value_.has_value()) {
+      this->parent_->perform_forced_calibration(value_.value());
+    }
+  }
+
+ protected:
+  TEMPLATABLE_VALUE(uint16_t, value)
+};
+
+template<typename... Ts> class FactoryResetAction : public Action<Ts...> {
```
Comment (jesserockz) [inline esphome/components/scd4x/automation.h]: You used `Parented` above, but not in this action?

**Review by jesserockz — APPROVED** [review]

File: esphome/components/scd4x/automation.h (hunk 3 above)
Comment (martgras) [reply esphome/components/scd4x/automation.h]: removed

**Pull request merged**